plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+    inline void setChannel(const char * _channel) { channel.set(_channel); }
+    inline const char * getChannel() { return channel.str(); }
+    inline void setMessage(const char * _message) { message.set(_message); }
+    inline const char * getMessage() { return message.str(); }
+    inline void wait(unsigned timeout) { msgSem.wait(timeout); }
+    inline void signal() { msgSem.signal(); }
+    inline void subActivated() { subActiveSem.signal(); }
+    inline void subActivationWait(unsigned timeout) { subActiveSem.wait(timeout); }
+    void subscribe(struct ev_loop * evLoop);
+    void unsubscribe();
+
+    void stopEvLoop()
+    {
+        CriticalBlock block(crit);
+        context->ev.delRead((void*)context->ev.data);
+        context->ev.delWrite((void*)context->ev.data);
+    }
+    //callback
+    static void subCB(redisAsyncContext * context, void * _reply, void * privdata);
+
+protected :
+    Semaphore msgSem;
+    Semaphore subActiveSem;
+    redisCallbackFn * callback;
+    StringAttr message;
+    StringAttr channel;
+};
+
+class SubscriptionThread : implements IThreaded, implements IInterface, public SubContainer
+{
+public :
+    SubscriptionThread(ICodeContext * ctx, const char * options, const char * channel, unsigned __int64 _database) : SubContainer(ctx, options, channel, _database), thread("SubscriptionThread", (IThreaded*)this)
+    {
+        evLoop = NULL;
+    }
+    virtual ~SubscriptionThread()
+    {
+        stopEvLoop();
+        wait(TIMEOUT);//give stopEvLoop time to complete
+        thread.stopped.signal();
+        thread.join();
+    }
+
+    void start()
+    {
+        thread.start();
+        subActivationWait(TIMEOUT);//wait for subscription to be acknowledged by redis
+    }
+
+    IMPLEMENT_IINTERFACE;
+
+private :
+    void main()
+    {
+        evLoop = ev_loop_new(0);
+        subscribe(evLoop);
+        signal();
+    }
+
+private :
+    CThreaded  thread;
+    struct ev_loop * evLoop;
+};
+SubContainer::SubContainer(ICodeContext * ctx, const char * options, const char * _channel, unsigned __int64 _database) : AsyncConnection(ctx, options, _database)
+{
+    channel.set(_channel);
+    callback = subCB;
+}
+SubContainer::SubContainer(ICodeContext * ctx, const char * options, const char * _channel, redisCallbackFn * _callback, unsigned __int64 _database) : AsyncConnection(ctx, options,_database )
+{
+    channel.set(_channel);
+    callback = _callback;
+}
+AsyncConnection::AsyncConnection(ICodeContext * ctx, const char * _options, unsigned __int64 _database) : Connection(ctx, _options, _database)
+{
+    createAndAssertConnection(ctx);
+    //could log server stats here, however async connections are not cached and therefore book keeping of only doing so for new servers may not be worth it.
+}
+void AsyncConnection::selectDb(ICodeContext * ctx)
+{
+    if (database == 0)
+        return;
+    attachLibev();
+    VStringBuffer cmd("SELECT %llu", database);
```

Comment:
I know this almost certainly doesn't work in windows, but we tend to avoid using ll and use " I64F " instead